### PR TITLE
Book recommender

### LINKED ISSUE #95: demo: book recommender
a RAG using OpenAI that recommends you books

we can use this dataset: https://www.kaggle.com/datasets/cristaliss/ultimate-book-collection-top-100-books-up-to-2023

### PR DESCRIPTION
**Pull request by neelasha23**
Closes #95 

----
📚 Documentation preview 📚: https://ploomber-doc--127.org.readthedocs.build/en/127/

### CODE REVIEW

**Comment by neelasha23** [comment]
Deployed here: https://small-rain-7598.ploomberapp.io/app @edublancas

File: examples/panel/book-recommender/app.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
+"""
+Demonstrates how to use the `ChatInterface` and a `callback` function to respond.
+
+The chatbot Assistant echoes back the message entered by the User.
+
+https://github.com/holoviz-topics/panel-chat-examples/blob/main/docs/examples/basics/basic_chat.py
+"""
```
Comment (edublancas) [inline examples/panel/book-recommender/app.py:7]: this docstring seems innacurate

File: examples/panel/book-recommender/app.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+        ],
+        seed=42,
+        n=1,
+    )
+    author = response.choices[0].message.content.upper()
+    return author if author in all_authors else ""
+
+
+def book_recommender_agent(user_query, verbose=False):
+    """An agent that can retrieve news by topic and summarizes them"""
```
Comment (edublancas) [inline examples/panel/book-recommender/app.py:47]: this docstring is innacurate

File: examples/panel/book-recommender/app.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+    titles = []
+    embeddings = []
+    for key, value in embeddings_json.items():
+        if value:
+            titles.append(key)
+            embeddings.append(value)
+    kdtree = KDTree(np.array(embeddings))
+    _, indexes = kdtree.query(store.get_one(user_query), k=min(len(titles), 3))
+
+    print(type(indexes))
```
Comment (edublancas) [inline examples/panel/book-recommender/app.py:74]: remove this print statements or use logging so we can disable/enable them easily

File: examples/panel/book-recommender/assets/author_to_title.pkl
Comment (edublancas) [inline examples/panel/book-recommender/assets/author_to_title.pkl:1]: delete this. I guess this was generated from the GoodReads data? if so, provide a script to generate this file

File: examples/panel/book-recommender/assets/title_to_description.pkl
Comment (edublancas) [inline examples/panel/book-recommender/assets/title_to_description.pkl:1]: delete this. I guess this was generated from the GoodReads data? if so, provide a script to generate this file

File: examples/panel/book-recommender/rag_book_recommender.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+import json
+import pickle
+import pandas as pd
+
+from pathlib import Path
+from openai import OpenAI
+
+client = OpenAI()
+
+file_path = os.path.join('assets', 'goodreads.csv')
```
Comment (edublancas) [inline examples/panel/book-recommender/rag_book_recommender.py:11]: there's no documentation on how to get this file, I'm assuming this was downloaded from kaggle, if so, mention it

File: examples/panel/book-recommender/rag_book_recommender.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
+import os
```
Comment (edublancas) [inline examples/panel/book-recommender/rag_book_recommender.py:1]: i'd recommend using pathlib.Path instead of os.path.join, it's cleaner

File: examples/panel/book-recommender/rag_book_recommender.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+from pathlib import Path
+from openai import OpenAI
+
+client = OpenAI()
+
+file_path = os.path.join('assets', 'goodreads.csv')
+df = pd.read_csv(file_path)
+
+
+file_path = os.path.join('assets', 'title_to_description.pkl')
```
Comment (edublancas) [inline examples/panel/book-recommender/rag_book_recommender.py:15]: do not use pickle files as serialization format. better to use JSON

File: examples/panel/book-recommender/rag_book_recommender.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+with open(file_path, 'rb') as file:
+    DESCRIPTIONS = pickle.load(file)
+
+
+def get_authors():
+    file_path = os.path.join('assets', 'author_to_title.pkl')
```
Comment (edublancas) [inline examples/panel/book-recommender/rag_book_recommender.py:21]: do not use pickle files as serialization format. better to use JSON

File: examples/panel/book-recommender/rag_book_recommender.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+    with open(file_path, 'rb') as file:
+        authors = pickle.load(file)
+    return authors
+
+
+def get_embeddings():
```
Comment (edublancas) [inline examples/panel/book-recommender/rag_book_recommender.py:27]: the overall code is a bit hard to follow because there are no docstrings. what's in the embeddings.json? I'm guessing the keys are the titles and the values are the embeddings produced from the description?

File: examples/panel/book-recommender/rag_book_recommender.py [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+    return embeddings_json
+
+
+def get_book_description_by_title(title):
+    return DESCRIPTIONS[title.upper()]
+
+
+class EmbeddingsStore:
+    def __init__(self):
+        self._path = Path("embeddings_another.json")
```
Comment (edublancas) [inline examples/panel/book-recommender/rag_book_recommender.py:40]: what's the difference between embeddings.json and embeddings_another.json?

File: examples/panel/book-recommender/rag_book_recommender.py [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+
+        if not self._path.exists():
+            self._data = {}
+        else:
+            self._data = json.loads(self._path.read_text())
+
+    def get_one(self, text, title=None):
+        if text in self._data:
+            return self._data[text]
```
Comment (edublancas) [inline examples/panel/book-recommender/rag_book_recommender.py:49]: I'm confused here

in line 83, you're doing:

```python
store.get_one(row["description"], row["title"])
```

I'm assuming description is a text describing the book and title is the book's title, this means that when get_one is executed, text=description...

then you check id the description is in `self._data`, and if so return the value (the embedding)

but then, in line 56, you do:  `self._data[title] = embedding`

meaning, you store the title as the key and the embedding as the value.

hence, line 48 will always be false, right?
```
        if text in self._data:
            return self._data[text]
```

because text (the description) is *never* in self._data, since you're storing the book's title, not the description

File: examples/panel/book-recommender/app.py (hunk 1 above)
Comment (edublancas) [reply examples/panel/book-recommender/app.py:7]: an overall description would help me better understand the cde:

- what the input data is and how it looks like
- how are embeddings generated
- what happens when a user submits a message

File: examples/panel/book-recommender/app.py [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+
+from rag_book_recommender import get_book_description_by_title, EmbeddingsStore, get_authors, get_embeddings
+
+
+client = OpenAI()
+
+pn.extension()
+
+store = EmbeddingsStore()
+all_authors = get_authors()
```
Comment (edublancas) [inline examples/panel/book-recommender/app.py:22]: this naming is confusing. I believe  get_authors returns an author -> book titles mapping, right? `all_authors` is innacurate

File: examples/panel/book-recommender/app.py [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+    )
+    author = response.choices[0].message.content.upper()
+    return author if author in all_authors else ""
+
+
+def book_recommender_agent(user_query, verbose=False):
+    """An agent that can retrieve news by topic and summarizes them"""
+    # determine the topic based on the query
+    embeddings_json = get_embeddings()
+    author = detect_author(user_query)
```
Comment (edublancas) [inline examples/panel/book-recommender/app.py:50]: This approach is pretty restrictive. if I understand correctly, whatever my input is, it'll be mapped to a single author (?) but users might ask something like "show me novels about spies", and there are many authors that would match that criteria.

instead. I suggest this:

- use openai's embedding AI to embed the user query
- then, find the closest matches to the embedding by looking into the book embeddings
- then retrieve the book's info (title, author, description), and display it

Comment (edublancas) [reply examples/panel/book-recommender/app.py:50]: update: I asked "novels about spies and got this"

image

which includes two authors, so I'm not sure what the logic is. the code suggests that it should only return books by a single author (?)

File: examples/panel/book-recommender/rag_book_recommender.py [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+        if self._path.exists():
+            self._path.unlink()
+            self._data = {}
+
+
+def compute_embeddings():
+    store = EmbeddingsStore()
+    store.clear()
+
+    for index, row in df.iterrows():
```
Comment (edublancas) [inline examples/panel/book-recommender/rag_book_recommender.py:81]: seems like you're computing embeddings for all books?

I suggest sampling (eg compute it for 50 books) until we agree on the whole logic

**Review by edublancas — CHANGES_REQUESTED** [review]
i have many questions about the logic

also, please ensure your docstrings are accurate and you either use logging or remove print statements before requesting a review

File: examples/panel/book-recommender/app.py (hunk 12 above)
Comment (neelasha23) [reply examples/panel/book-recommender/app.py:50]: No it doesn't return books by a single author. Since author is a separate column if user asks recommend books by `some-author` it doesn't work if we only compute similarity with the description field. So if any specific author is mentioned it first filters based on authors and then returns. If no author is mentioned it compares with embeddings of all available books

File: examples/panel/book-recommender/app.py (hunk 2 above)
Comment (neelasha23) [reply examples/panel/book-recommender/app.py]: [fixed](https://github.com/ploomber/doc/pull/127/files#diff-99d50d468b15d576674758b1fbde220b390f380dc2398dd8b4024a04218dd2ccR86-R87)

File: examples/panel/book-recommender/assets/author_to_title.pkl
Comment (neelasha23) [reply examples/panel/book-recommender/assets/author_to_title.pkl:1]: [fixed](https://github.com/ploomber/doc/pull/127/commits/ebfd613fefc74af6ad51eeba43df786e893d38c1#)

File: examples/panel/book-recommender/assets/title_to_description.pkl
Comment (neelasha23) [reply examples/panel/book-recommender/assets/title_to_description.pkl:1]: [fixed](https://github.com/ploomber/doc/pull/127/commits/ebfd613fefc74af6ad51eeba43df786e893d38c1#)
[Added generation script](https://github.com/ploomber/doc/pull/127/files#diff-396e0215830b7ad98dc178d3fd362e3e0a577ab3509a784b915a8242e092be98)

File: examples/panel/book-recommender/rag_book_recommender.py (hunk 4 above)
Comment (neelasha23) [reply examples/panel/book-recommender/rag_book_recommender.py]: [fixed](https://github.com/ploomber/doc/pull/127/files#diff-396e0215830b7ad98dc178d3fd362e3e0a577ab3509a784b915a8242e092be98R9-R11)

File: examples/panel/book-recommender/rag_book_recommender.py (hunk 5 above)
Comment (neelasha23) [reply examples/panel/book-recommender/rag_book_recommender.py]: [fixed](https://github.com/ploomber/doc/pull/127/files#diff-396e0215830b7ad98dc178d3fd362e3e0a577ab3509a784b915a8242e092be98R40-R41)

File: examples/panel/book-recommender/rag_book_recommender.py (hunk 6 above)
Comment (neelasha23) [reply examples/panel/book-recommender/rag_book_recommender.py]: [Using json now](https://github.com/ploomber/doc/pull/127/files#diff-396e0215830b7ad98dc178d3fd362e3e0a577ab3509a784b915a8242e092be98R38-R50)

File: examples/panel/book-recommender/rag_book_recommender.py (hunk 7 above)
Comment (neelasha23) [reply examples/panel/book-recommender/rag_book_recommender.py]: [Using json now](https://github.com/ploomber/doc/pull/127/files#diff-396e0215830b7ad98dc178d3fd362e3e0a577ab3509a784b915a8242e092be98R38-R50)

File: examples/panel/book-recommender/rag_book_recommender.py (hunk 8 above)
Comment (neelasha23) [reply examples/panel/book-recommender/rag_book_recommender.py]: [fixed](https://github.com/ploomber/doc/pull/127/files#diff-99d50d468b15d576674758b1fbde220b390f380dc2398dd8b4024a04218dd2ccR36-R38)

File: examples/panel/book-recommender/rag_book_recommender.py (hunk 9 above)
Comment (neelasha23) [reply examples/panel/book-recommender/rag_book_recommender.py]: Added ` embeddings_another.json` by mistake

File: examples/panel/book-recommender/rag_book_recommender.py (hunk 10 above)
Comment (neelasha23) [reply examples/panel/book-recommender/rag_book_recommender.py]: I have replaced this logic with [this function](https://github.com/ploomber/doc/pull/127/files#diff-396e0215830b7ad98dc178d3fd362e3e0a577ab3509a784b915a8242e092be98R16-R34) since the `EmbeddingsStore` is being used for caching embeddings for dynamically generated embeddings. In this use-case we already have a static book dataset so I have pre-computed all the embeddings and stored in the file, so I don't think we need the caching mechanism? Is this assumption correct?

File: examples/panel/book-recommender/app.py (hunk 11 above)
Comment (neelasha23) [reply examples/panel/book-recommender/app.py]: [fixed](https://github.com/ploomber/doc/pull/127/files#diff-99d50d468b15d576674758b1fbde220b390f380dc2398dd8b4024a04218dd2ccR31-R32)

File: examples/panel/book-recommender/app.py (hunk 12 above)
Comment (neelasha23) [reply examples/panel/book-recommender/app.py:91]: Added a [comment](https://github.com/ploomber/doc/pull/127/files#diff-99d50d468b15d576674758b1fbde220b390f380dc2398dd8b4024a04218dd2ccR89-R90) for clarity

File: examples/panel/book-recommender/rag_book_recommender.py (hunk 13 above)
Comment (neelasha23) [reply examples/panel/book-recommender/rag_book_recommender.py]: Have modified the `embeddings.json` to include 50 embeddings. But does it matter because these are being pre-computed anyways?

**Comment by neelasha23** [comment]
App with 50 embeddings: https://blue-mouse-7632.ploomberapp.io/app @edublancas

File: examples/panel/book-recommender/rag_book_recommender.py (hunk 4 above)
Comment (edublancas) [reply examples/panel/book-recommender/rag_book_recommender.py]: Is this mentioned in the readme? it'd better to document this in the readme than the code

File: examples/panel/book-recommender/rag_book_recommender.py (hunk 10 above)
Comment (edublancas) [reply examples/panel/book-recommender/rag_book_recommender.py]: yeah we can get rid of the EmbeddingsStor if you want

File: examples/panel/book-recommender/assets/embeddings.json
Comment (edublancas) [inline examples/panel/book-recommender/assets/embeddings.json:1]: don't upload data files to git

File: examples/panel/book-recommender/assets/goodreads.csv
Comment (edublancas) [inline examples/panel/book-recommender/assets/goodreads.csv:1]: don't upload data files to git

File: examples/panel/book-recommender/requirements.txt [hunk 14]
```diff
@@ -0,0 +1,4 @@
+panel
```
Comment (edublancas) [inline examples/panel/book-recommender/requirements.txt:1]: also add a requirements.lock.txt, generate it with `pip freeze`

**Review by edublancas — CHANGES_REQUESTED** [review]
minor comments, looking good!

File: examples/panel/book-recommender/rag_book_recommender.py (hunk 10 above)
Comment (neelasha23) [reply examples/panel/book-recommender/rag_book_recommender.py]: Removed it already since now embeddings are being generated using [this function](https://github.com/ploomber/doc/pull/127/files#diff-396e0215830b7ad98dc178d3fd362e3e0a577ab3509a784b915a8242e092be98R16-R34)

File: examples/panel/book-recommender/rag_book_recommender.py (hunk 4 above)
Comment (neelasha23) [reply examples/panel/book-recommender/rag_book_recommender.py]: [added](https://github.com/ploomber/doc/pull/127/files#diff-3b853450dda7ca635422b6a6d2c988f70e93aea79c8f8878384d0fa175e013e0R13-R16)

File: examples/panel/book-recommender/assets/embeddings.json
Comment (neelasha23) [reply examples/panel/book-recommender/assets/embeddings.json:1]: [removed assets folder](https://github.com/ploomber/doc/pull/127/commits/bfee1d5bd4b2e653f8aaddccfe8cee82f8ddd195)

File: examples/panel/book-recommender/requirements.txt (hunk 14 above)
Comment (neelasha23) [reply examples/panel/book-recommender/requirements.txt:1]: [added](https://github.com/ploomber/doc/pull/127/commits/a6d5d38d9db668b88244a7c2f31b5520c91d0a1c)

**Comment by neelasha23** [comment]
Addressed comment. Please review @edublancas

**Review by edublancas — APPROVED** [review]

**Comment by edublancas** [comment]
@neelasha23 please delete the application with 50 embeddings

ensure the other app is running the latest version and prepare the social media posts!
